cmd/oras/push.go

```diff
@@ -63,76 +72,62 @@ Example - Push file to the HTTP registry:
 		},
 	}
 
-	cmd.Flags().StringVarP(&opts.manifestConfigRef, "manifest-config", "", "", "manifest config file")
-	cmd.Flags().StringVarP(&opts.manifestAnnotations, "manifest-annotations", "", "", "manifest annotation file")
-	cmd.Flags().BoolVarP(&opts.pathValidationDisabled, "disable-path-validation", "", false, "skip path validation")
-	cmd.Flags().BoolVarP(&opts.verbose, "verbose", "v", false, "verbose output")
-	cmd.Flags().BoolVarP(&opts.debug, "debug", "d", false, "debug mode")
-	cmd.Flags().StringArrayVarP(&opts.configs, "config", "c", nil, "auth config path")
-	cmd.Flags().StringVarP(&opts.username, "username", "u", "", "registry username")
-	cmd.Flags().StringVarP(&opts.password, "password", "p", "", "registry password")
-	cmd.Flags().BoolVarP(&opts.insecure, "insecure", "", false, "allow connections to SSL registry without certs")
-	cmd.Flags().BoolVarP(&opts.plainHTTP, "plain-http", "", false, "use plain http and not https")
+	option.ApplyFlags(&opts, cmd.Flags())
 	return cmd
 }
 
 func runPush(opts pushOptions) error {
-	ctx := context.Background()
-	if opts.debug {
-		logrus.SetLevel(logrus.DebugLevel)
-	} else if !opts.verbose {
-		ctx = ctxo.WithLoggerDiscarded(ctx)
-	}
+	ctx, _ := opts.SetLoggerLevel()
 
-	// load files
-	var (
-		annotations map[string]map[string]string
-		store       = content.NewFileStore("")
-		pushOpts    []oras.PushOpt
-	)
-	defer store.Close()
-	if opts.manifestAnnotations != "" {
-		if err := decodeJSON(opts.manifestAnnotations, &annotations); err != nil {
-			return err
-		}
-		if value, ok := annotations[annotationConfig]; ok {
-			pushOpts = append(pushOpts, oras.WithConfigAnnotations(value))
-		}
-		if value, ok := annotations[annotationManifest]; ok {
-			pushOpts = append(pushOpts, oras.WithManifestAnnotations(value))
-		}
+	ref, err := registry.ParseReference(opts.targetRef)
+	if err != nil {
+		return err
 	}
-	if opts.manifestConfigRef != "" {
-		filename, mediaType := parseFileRef(opts.manifestConfigRef, artifact.UnknownConfigMediaType)
-		file, err := store.Add(annotationConfig, mediaType, filename)
-		if err != nil {
+	reg, err := opts.NewRegistry(ref.Registry, opts.Common)
+	if err != nil {
+		return err
+	}
+	dst, err := reg.Repository(ctx, ref.Repository)
+	if err != nil {
+		return err
+	}
+
+	// Load annotations
+	var annotations map[string]map[string]string
+	if opts.ManifestAnnotations != "" {
+		if err := decodeJSON(opts.ManifestAnnotations, &annotations); err != nil {
 			return err
 		}
-		file.Annotations = nil
-		pushOpts = append(pushOpts, oras.WithConfig(file))
 	}
-	if opts.pathValidationDisabled {
-		pushOpts = append(pushOpts, oras.WithNameValidation(nil))
-	}
-	files, err := loadFiles(store, annotations, &opts)
+
+	// Prepare manifest
+	store := file.New("")
+	defer store.Close()
+	store.AllowPathTraversalOnWrite = opts.PathValidationDisabled
+
+	// Ready to push
+	tracker := status.NewPushTracker(dst, opts.Verbose)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```

Comment:
What's the purpose of having a cancellable context? I'm not sure about the intention here.